pkg/controller/useraccount/useraccount_controller.go

```diff
@@ -239,6 +248,56 @@ func (r *ReconcileUserAccount) ensureIdentity(logger logr.Logger, userAcc *toolc
 	return identity, false, nil
 }
 
+func (r *ReconcileUserAccount) ensureNSTemplateSet(logger logr.Logger, userAcc *toolchainv1alpha1.UserAccount) (*toolchainv1alpha1.NSTemplateSet, bool, error) {
+	name := userAcc.Name
+
+	// create if not found
+	nsTmplSet := &toolchainv1alpha1.NSTemplateSet{}
+	if err := r.client.Get(context.TODO(), types.NamespacedName{Name: userAcc.Name, Namespace: userAcc.Namespace}, nsTmplSet); err != nil {
+		if errors.IsNotFound(err) {
+			logger.Info("creating a new NSTemplateSet", "name", name)
+			if err := r.setStatusProvisioning(userAcc); err != nil {
+				return nil, false, err
+			}
+			nsTmplSet = newNSTemplateSet(userAcc)
+			if err := controllerutil.SetControllerReference(userAcc, nsTmplSet, r.scheme); err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err,
+					"failed to set controller reference for NSTemplateSet '%s'", name)
+			}
+			if err := r.client.Create(context.TODO(), nsTmplSet); err != nil {
+				return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err,
+					"failed to create NSTemplateSet '%s'", name)
+			}
+			logger.Info("NSTemplateSet created successfully", "name", name)
+			return nsTmplSet, true, nil
+		}
+		return nil, false, r.wrapErrorWithStatusUpdate(logger, userAcc, r.setStatusNSTemplateSetCreationFailed, err, "failed to get NSTemplateSet '%s'", name)
+	}
+
+	// update if not same
+	equal := compareNSTemplateSet(nsTmplSet.Spec, userAcc.Spec.NSTemplateSet)
```

Comment:
As I can see, there is `DeepCopy` [link](https://github.com/codeready-toolchain/api/blob/master/pkg/apis/toolchain/v1alpha1/zz_generated.deepcopy.go) but there is NO `DeepEqual`. Let me know if I missing any thing here.

Reply:
I meant `reflect.DeepEqual` from Golang

Reply:
`reflect.DeepEqual` will NOT work when we have same namespaces in both but in different order.

Reply:
Sure, that is correct. But do we expect that it could be in a different order? You assign the NsTemplateSet from the UserAcc without any changes, so I guess that it should be equal. The  NsTemplateSet.spec is managed by this controller, so there shouldn't be any changes in the order...
I made this assumption in the MUR <-> UserAcc already and I guess that it is fine. 
My point is that I don't see any reason why we should bring additional complexity to the code.

Reply:
It's good to make things robust here.  If we get namespace out of order then it's difficult to debug such issue.  If you are fine with either then I would like to keep as it is.

Reply:
Ok, but then we should introduce the same logic to the MUR <-> UserAcc synchronization - to keep the logic consistent. 
Could you please move the comparison logic to either common or api repo, so it can be reused in both operators? The api repo would make maybe more sense as it would be close to the types that it compares, WDYT?

Reply:
@MatousJobanek yes, we should move it to `api`, I will do it :+1: